mne/decoding/search_light.py

```diff
@@ -0,0 +1,332 @@
+            Transformed values generated by each estimator.
+        """
+        method = 'transform'
+        if not hasattr(self.base_estimator, 'transform'):
+            method = 'predict'
+        self._transform(X, method)
+
+    def predict(self, X):
+        """Predict the dataset with a series of independent estimators.
+
+        Parameters
+        ----------
+        X : array, shape(n_samples, n_features, n_iteration)
+            The training input samples. For each iteration, a clone estimator
+            is fitted independently.
+
+        Returns
+        -------
+        y_pred : array, shape(n_samples, n_train_iterations, n_test_iterations)
+            Predicted values for each estimator.
+        """
+        return self._transform(X, 'predict')
+
+    def predict_proba(self, X):
+        if not hasattr(self.base_estimator, 'predict_proba'):
+            ValueError('self.base_estimator does not have `predict_proba` '
+                       'method.')
+        return self._transform(X, 'predict_proba')
+
+    def decision_function(self, X):
+        if not hasattr(self.base_estimator, 'decision_function'):
+            ValueError('self.base_estimator does not have `decision_function` '
+                       'method.')
+        return self._transform(X, 'decision_function')
+
+    def score(self, X, y):
+        # FIXME: not sure how to design
+        raise NotImplementedError()
+
+
+def _gl_transform(estimators, X, method):
+    """Transform the dataset by applying each estimator to all iteration of
+    the data.
+
+    Parameters
+    ----------
+    X : array, shape(n_samples, n_features, n_iteration)
+        The training input samples. For each iteration, a clone estimator
+        is fitted independently.
+
+    Returns
+    -------
+    Xt : array, shape(n_samples, n_iterations)
+        Transformed values generated by each estimator.
+    """
+    n_sample, n_chan, n_iter = X.shape
+    for ii, est in enumerate(estimators):
+        # stack generalized data for faster prediction
+        X_stack = np.transpose(X, [1, 0, 2])
+        X_stack = np.reshape(X_stack, [n_chan, n_sample * n_iter]).T
+        transform = getattr(est, method)
+        _y_pred = transform(X_stack)
+        # unstack generalizations
+        if _y_pred.ndim == 2:
+            _y_pred = np.reshape(_y_pred, [n_sample, n_iter])
+        else:
+            shape = np.r_[n_sample, n_iter, _y_pred.shape[1:]]
+            _y_pred = np.reshape(_y_pred, shape)
+        # init
+        if ii == 0:
+            y_pred = _gl_init_pred(_y_pred, X, len(estimators))
+        y_pred[:, ii, ...] = _y_pred
+    return y_pred
+
+
+def _gl_init_pred(y_pred, X, n_train):
+    n_sample, n_chan, n_iter = X.shape
+    if y_pred.ndim == 3:
+        y_pred = np.zeros((n_sample, n_train, n_iter, y_pred.shape[-1]))
+    else:
+        y_pred = np.zeros((n_sample, n_train, n_iter))
+    return y_pred
```

Comment:
Ok thanks for clarifying, this is an excellent idea. Good thing is that the
nightmare will then be over :)

On Sat, Jul 2, 2016 at 7:51 PM Jean-Rémi KING <notifications@github.com>
wrote:

> In mne/decoding/search_light.py
> <https://github.com/mne-tools/mne-python/pull/3381#discussion_r69380900>:
>
> > +            shape = np.r_[n_sample, n_iter, _y_pred.shape[1:]]
> > +            _y_pred = np.reshape(_y_pred, shape)
> > +        # init
> > +        if ii == 0:
> > +            y_pred = _gl_init_pred(_y_pred, X, len(estimators))
> > +        y_pred[:, ii, ...] = _y_pred
> > +    return y_pred
> > +
> > +
> > +def _gl_init_pred(y_pred, X, n_train):
> > +    n_sample, n_chan, n_iter = X.shape
> > +    if y_pred.ndim == 3:
> > +        y_pred = np.zeros((n_sample, n_train, n_iter, y_pred.shape[-1]))
> > +    else:
> > +        y_pred = np.zeros((n_sample, n_train, n_iter))
> > +    return y_pred
>
> No conceptual difference;
>
> The main motivation is
>
>    - to make these searchlight objects compatible with sklearn pipeline
>    i.e. cross_val_score(Xdawn(), SearchLight())
>    - to make search light across other features than time (this is what I
>    used I my last submission for time freq analyses)
>
> Refactoring the GAT directly is going a nightmare so I prefer starting
> from a clean base, and then do something like
>
> class GeneralizationAcrossTime(many_params):
>    def fit(epochs, y=None):
>         picks = pick_channels(epochs.info, meg=True, eeg=True ...)
>         X = epochs.pick_types(picks).get_data()
>         y = y if y is not None else epochs.events[:, 2]
>         self._searchlight.fit(X, y)
>        return self
>
> it'll be much easier and simple: i.e. have smart objects that deals with
> mne objects for default users, and have more abstract objects for more
> advanced users.
>
> Furthermore, I checked that
>
>    - this code is equally CPU/memory efficient to the current GAT
>    - dealing with various predict method (predict_proba,
>    'decision_functionetc) can be dealt at the appropriate level e.g.SearchLight(substitute_predict(LogisticRegression(),
>    method='predict_proba))
>    - it will remain compatible with estimators based on multiple
>    time-sample by simply doing: make_pipeline(TimeEmbedder(10),
>    SearchLight()), although this is slightly less memory efficient..
>    #NoFreeLunch. Perhaps we'll find a way to deal with masks instead.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3381/files/02b9b7688bd19d76853acd3864153dfe3eccdfb7#r69380900>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AB0fijZNwhHUWIpt7og8nHhHQJlgH-5uks5qRqUQgaJpZM4JDsU1>
> .
>